main-board/src/sd_interface.rs

```diff
@@ -0,0 +1,152 @@
+
+use core::{marker::PhantomData};
+
+use defmt::{warn, info};
+
+use crate::{
+    hal,
+    pac,
+    sd,
+};
+
+pub struct TimeSink {
+    _marker: PhantomData<*const ()>,
+}
+
+impl TimeSink { 
+    fn new() -> Self {
+        TimeSink {
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl sd::TimeSource for TimeSink {
+    fn get_timestamp(&self) -> sd::Timestamp {
+        sd::Timestamp {
+            year_since_1970: 0,
+            zero_indexed_month: 0,
+            zero_indexed_day: 0,
+            hours: 0,
+            minutes: 0,
+            seconds: 0,
+        }
+    }
+}
+
+pub struct SdInterface
+{
+    pub sd_controller: sd::Controller<sd::SdMmcSpi<hal::sercom::spi::Spi<hal::sercom::spi::Config<hal::sercom::spi::Pads<hal::sercom::Sercom1, hal::sercom::IoSet1, hal::gpio::Pin<hal::gpio::PA19, hal::gpio::Alternate<hal::gpio::C>>, hal::gpio::Pin<hal::gpio::PA16, hal::gpio::Alternate<hal::gpio::C>>, hal::gpio::Pin<hal::gpio::PA17, hal::gpio::Alternate<hal::gpio::C>>>>, hal::sercom::spi::Duplex>, hal::gpio::Pin<hal::gpio::PA18, hal::gpio::Output<hal::gpio::PushPull>>>, TimeSink>,
```

Comment:
This type is too long and specific. In short, this is not supper flexible, only being good for a specific set of pins on our specific board. Ideally, `SdInterface` would be useable with any `Constroller`. So, time for generics! I'll be fair, this is a bit complicated to do. I tried to get something this morning, but didn't have time to get it to work. However, here is how your struct should look:
```rust
pub struct SdInterface<D, T>
where
    D: BlockDevice,
    T: TimeSource,
{
    pub sd_controller: Controller<D, T>,
    pub volume: Volume,
    pub root_directory: Directory,
    pub file: File,
}
```